### PR TITLE
Support manual signaling of channel readiness

### PR DESCRIPTION
**Pull request by johncantrell97**
Second pass at implementing https://github.com/lightningdevkit/rust-lightning/issues/2106

- Refactors the unnamed tuple that gets returned as part of `do_chain_work` flow with a `ChainActionUpdates` struct.

- Adds a manually_signal_channel_ready config option to `ChannelHandshakeConfig` for users to configure a channel with this option.  Defaults to false.

- Saves this option to Channel using `requires_manual_readiness_signal` and this same flag is updated when user manually signals readiness.

- Tracks emission of `PendingChannelReady` event using new flag on `Channel` `pending_channel_ready_event_emitted` similar to how the `ChannelReady` event works.

- Adds a new `ChainActionUpdates` field to signal to ChannelManager when it's time to generate the `PendingChannelReady` event.

### CODE REVIEW

**Comment by johncantrell97** [comment]
@wpaulino would love another round of review on this approach when you get a chance

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -6484,6 +6500,7 @@ impl<Signer: WriteableEcdsaChannelSigner> Writeable for Channel<Signer> {
 			(23, channel_ready_event_emitted, option),
 			(25, user_id_high_opt, option),
 			(27, self.channel_keys_id, required),
+			(29, requires_manual_readiness_signal, option)
```
[the remark this answers is not in the document]
Comment (johncantrell97) [reply lightning/src/ln/channel.rs]: Interesting.  So to make sure I understand.  It's a user who had enabled this feature and got partially through the channel open process.  They then downgrade LDK to a version before this feature exists.  It means this TLV is persisted and the old version will just ignore it because it doesn't know about it.

In the downgraded ldk it will automatically signal channel ready without asking them.

I don't really think that's a problem? They explicitly downgraded to a version of LDK that no longer supports manual signaling?

If we make this even then it means when they try to start up the downgraded LDK it will panic when it tries to deserialize this type it doesn't know about?  

Basically forcing them to upgrade or wait for a bugfix for whatever reason they downgraded in the first place?

If all of this is correct then I guess I'm not so sure what is the better experience.  I guess it really just depends on _why_ they want to manually signal and partially why they are downgrading.

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -4389,6 +4406,42 @@ where
...
+		let peer_state = &mut *peer_state_lock;
+		let pending_msg_events = &mut peer_state.pending_msg_events;
+		match peer_state.channel_by_id.entry(channel_id.clone()) {
+			hash_map::Entry::Occupied(mut channel) => {
+				let best_block_height = self.best_block.read().unwrap().height();
+				let channel = channel.get_mut();
+				let (channel_ready_opt, _) = channel.check_get_channel_ready(best_block_height);
+				match channel_ready_opt {
+					Some(channel_ready) => {
+						channel.unset_requires_manual_readiness_signal();
```
[the remark this answers is not in the document]
Comment (johncantrell97) [reply lightning/src/ln/channelmanager.rs:4431]: Hm yeah, that's how I originally had it but this comment https://github.com/lightningdevkit/rust-lightning/pull/2109#discussion_r1142544337 made it seem like we wanted to re-generate the PendingChannelReady event on restart if they hadn't signaled yet.

So the flow is that a chain event happens, we generate PendingChannelReady, flip `requires_manual_readiness_signal` to false, then LDK terminates.

On next startup, assuming the flag update was persisted, we will go ahead and automatically send ChannelReady when a block is connected without the user manually signaling it's okay.

So if we only have one flag and don't flip it until they signal then every time a block is found we would keep generating the event.

So one flag tracks whether or not we already emitted the event to prevent duplicates on subsequent chain events and the other is supposed to track whether this channel _still_ requires manual signal from user.

I guess the way I have it currently will also persist the "we emitted" flag and would prevent a re-emission on startup which is why Matt mentioned in-memory only?  

Thoughts? I'm kind of confusing myself again around this logic.

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -2461,7 +2467,7 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 
 		log_info!(logger, "Received funding_signed from peer for channel {}", log_bytes!(self.channel_id()));
 
-		let need_channel_ready = self.check_get_channel_ready(0).is_some();
+		let need_channel_ready = self.check_get_channel_ready(0).0.is_some();
```
[the remark this answers is not in the document]
Comment (johncantrell97) [reply lightning/src/ln/channel.rs]: Only in the case the user has manually signal enabled, right? The logic should be unchanged if `requires_manual_readiness_signal` is false, no?

**Comment by johncantrell97** [comment]
Ok so is the latest thinking to:

- remove the PendingChannelReady event entirely
- make `signal_channel_readiness` used to trigger `ChannelReady` message and state regardless of config/set up (assuming funding_signed has already been sent) so that people can "opt in" to any-conf they want?
- and then the manual signal flag just stops the automatic sending of ChannelReady but the end-user should use `signal_channel_readiness` whenever they want _after_ receiving the ChannelPending event?

i guess for inbound channels (what we're concerned with here) we wont necessarily have the funding tx in our bitcoind when we receive ChannelPending where as with this PendingChannelReady we will have received confirmations already assuming we aren't using 0-conf inbound (which we wont).

it means our workflow would have to be something like maintain a queue of channel_ids / funding_txo that we push to when we receive a `ChannelPending` and then add another object that implements `Listen` and when we finally see it in a block we can trigger readiness if it passes our screen?

Or is there a simpler workflow we can use? or am I misunderstanding the suggestion?

**Comment by johncantrell97** [comment]
> Yea, that's correct. If that's substantial additional complexity on your end we can do the event still or make the event appear on 1-conf irrespective of the channel config. We could also just make the event repeat, but I think that's a bit weird.

Ok, that's not too bad.  It's much better than one option I had floated to avoid changing ldk all which was to basically filter and screen every single tx before we feed it to ldk.  This allows us to only screen relevant txs.

Actually, couldn't we basically do this without any of these changes by just querying our channels on each block connected to see if there are any channels that are (funding_txo.is_some() && !is_outbound  && confirmations < confirmations_required).

for each channel in that state we can check if the funding tx is in the block, if it is, screen the inputs and close the channel id needed.

as long as we don't accept 0 (or 1) conf inbound channels this should be fine?

**Comment by johncantrell97** [comment]
Ah, well i'm not actually doing it _before_ providing to LDK.  I guess I could change that but currently it's just adding a third `Listen` implementation to the SpvClient (ChannelManager, ChainMonitor, and this new ChannelScreener).  

I guess maybe it's better to have a single Listener that can pipe them wherever/whenever it needs.

As for this PR, I think it's probably not so important for us anymore.  Though having manual signaling would be much more explicit then this new approach.

Do you think it's worth implementing this feature without the event?

**Comment by ZmnSCPxj-jr** [comment]
@johncantrell97 Looks like this may be useful for a 0-conf channel open flow from a swap-in-potentiam address, what can be done to revive this PR?

**Comment by ZmnSCPxj-jr** [comment]
In addition, I need to get the `scriptPubKey` of the funding outpoint.
